### PR TITLE
Add flag to the tls config that determines to use the system cert pool or an empty pool

### PR DESCRIPTION
**Pull request by cyli**
This is a stab at fixing https://github.com/docker/docker/issues/30450#issuecomment-278852038.

Note:  I'm not attached to the name of the flag - I'm very bad at naming things.  I just wanted to name it something does not change existing behavior if set to `false` (zero value).

@dmcgowan @nathanleclaire @diogomonica

### CODE REVIEW

File: tlsconfig/config_test.go [hunk 1]
```diff
@@ -240,11 +277,105 @@ func TestConfigServerTLSClientCASet(t *testing.T) {
 	if tlsConfig.ClientAuth != tls.VerifyClientCertIfGiven {
 		t.Fatal("ClientAuth was not set to what was in the options")
 	}
-	if tlsConfig.ClientCAs == nil || len(tlsConfig.ClientCAs.Subjects()) != 2 {
+	basePool, err := SystemCertPool()
```
Comment (cyli) [inline tlsconfig/config_test.go]: Note that this change is unrelated to this PR - these tests just didn't seem to pass on mac os previously.

File: tlsconfig/config_test.go [hunk 2]
```diff
@@ -240,11 +277,105 @@ func TestConfigServerTLSClientCASet(t *testing.T) {
...
+	if err != nil {
+		basePool = x509.NewCertPool()
+	}
+	if tlsConfig.ClientCAs == nil || len(tlsConfig.ClientCAs.Subjects()) != len(basePool.Subjects())+2 {
```
Comment (diogomonica) [inline tlsconfig/config_test.go]: maybe add a comment on why we're doing `basePool.Subjects +2`

**Comment by diogomonica** [comment]
LGTM

File: tlsconfig/config.go [hunk 3]
```diff
@@ -66,11 +71,17 @@ func ClientDefault() *tls.Config {
...
 
 // certPool returns an X.509 certificate pool from `caFile`, the certificate file.
-func certPool(caFile string) (*x509.CertPool, error) {
+func certPool(caFile string, exclusivePool bool) (*x509.CertPool, error) {
 	// If we should verify the server, we need to load a trusted ca
-	certPool, err := SystemCertPool()
-	if err != nil {
-		return nil, fmt.Errorf("failed to read system certificates: %v", err)
+	var (
+		certPool = x509.NewCertPool()
```
Comment (dmcgowan) [inline tlsconfig/config.go]: `x509.NewCertPool()` can just be called in the else case to avoid creating 2 maps when the system pool is used.

File: tlsconfig/config_test.go [hunk 4]
```diff
@@ -46,7 +83,7 @@ func generateCertificate(t *testing.T, signer crypto.Signer, out io.Writer) {
 	}
 }
 
-// generates a multiple-certificate file with both RSA and ECDSA certs and
+// generates a multiple-certificate CA files with both RSA and ECDSA certs and
```
Comment (dmcgowan) [inline tlsconfig/config_test.go]: files should not be plural, `CA file`

**Comment by nathanleclaire** [comment]
This is intended to be a CLI flag? Or am I missing something here?

**Comment by cyli** [comment]
@nathanleclaire Ah sorry, I wasn't very clear in the description.   The intention was to address https://github.com/docker/docker/issues/30450#issuecomment-279479503.

In https://github.com/docker/docker/blob/master/cli/command/cli.go#L246, where we create the client config to communicate with the daemon, we can pass an option to enable exclusive pools, since if a client has a CA configured for a particular remote daemon, it seems superfluous to also trust the system certs.

Depending on the use case, the user may want to be able to specify whether they want to use an exclusive pool of configuring a client or server, or trust the system pool + some other certs.

**Comment by nathanleclaire** [comment]
Ah, OK, thanks, that sounds good to me.

**Comment by dmcgowan** [comment]
LGTM

**Comment by cyli** [comment]
(I also cannot merge :))

**Comment by cyli** [comment]
@dmcgowan Thank you!
